# HG changeset patch
# User Piotr Held <pjheld@gmail.com>
# Date 1505946873 25200
#      Wed Sep 20 15:34:33 2017 -0700
# Node ID f8b8a8e6bd7939664f7d4126641b4c554227e93e
# Parent  0d7a7fc657ff6083edb1d8f9bc07fec3dfa28b03
make local functions and method names unique for parser (Bug #52075)

* oct-parse.in.yy (fcn_name): Make ID unique by prepending
base_lexer::fcn_file_name + ">" to all local subfunctions before
calling parent_scope_info::name_current_scope.

* test/bug-52075/bug-52075.tst, test/bug-52075/class_bug52075.m,
test/bug-52075/module.mk: New files.

* test/module.mk: Update.

diff --git a/libinterp/parse-tree/oct-parse.in.yy b/libinterp/parse-tree/oct-parse.in.yy
--- a/libinterp/parse-tree/oct-parse.in.yy
+++ b/libinterp/parse-tree/oct-parse.in.yy
@@ -1508,16 +1508,23 @@ function_beg    : push_fcn_symtab FCN
                       lexer.maybe_classdef_get_set_method = true;
                   }
                 ;
 
 fcn_name        : identifier
                   {
                     std::string id = $1->name ();
 
+                    // Make classdef local functions unique from
+                    // classdef methods.
+
+                    if (parser.m_parsing_local_functions
+                        && parser.m_curr_fcn_depth == 1)
+                      id = lexer.fcn_file_name + ">" + id;
+
                     if (! parser.m_function_scopes.name_current_scope (id))
                       {
                         parser.bison_error ("duplicate subfunction or nested function name",
                                             $1->line (), $1->column ());
 
                         delete $1;
 
                         YYABORT;
diff --git a/test/bug-52075/bug-52075.tst b/test/bug-52075/bug-52075.tst
new file mode 100644
--- /dev/null
+++ b/test/bug-52075/bug-52075.tst
@@ -0,0 +1,22 @@
+## Copyright (C) 2015-2017 Piotr Held
+##
+## This file is part of Octave.
+##
+## Octave is free software; you can redistribute it and/or modify it
+## under the terms of the GNU General Public License as published by
+## the Free Software Foundation; either version 3 of the License, or
+## (at your option) any later version.
+##
+## Octave is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Octave; see the file COPYING.  If not, see
+## <http://www.gnu.org/licenses/>.
+
+%!test
+%! a = class_bug52075 ();
+%! assert (a.duplicate, 'method');
+%! assert (a.call_local_dup, 'local_foo');
diff --git a/test/bug-52075/class_bug52075.m b/test/bug-52075/class_bug52075.m
new file mode 100644
--- /dev/null
+++ b/test/bug-52075/class_bug52075.m
@@ -0,0 +1,14 @@
+classdef class_bug52075
+  methods
+    function ret = duplicate (self)
+      ret = 'method';
+    end
+    function ret = call_local_dup (self)
+      ret = duplicate ();
+    end
+  end
+end
+
+function ret = duplicate
+  ret = 'local_foo';
+end
diff --git a/test/bug-52075/module.mk b/test/bug-52075/module.mk
new file mode 100644
--- /dev/null
+++ b/test/bug-52075/module.mk
@@ -0,0 +1,5 @@
+bug_52075_TEST_FILES = \
+  %reldir%/bug-52075.tst \
+  %reldir%/class_bug52075.m
+
+TEST_FILES += $(bug_52075_TEST_FILES)
diff --git a/test/module.mk b/test/module.mk
--- a/test/module.mk
+++ b/test/module.mk
@@ -56,16 +56,17 @@ include %reldir%/bug-46660/module.mk
 include %reldir%/bug-49379/module.mk
 include %reldir%/bug-50014/module.mk
 include %reldir%/bug-50035/module.mk
 include %reldir%/bug-50716/module.mk
 include %reldir%/bug-51192/module.mk
 include %reldir%/bug-51532/module.mk
 include %reldir%/bug-51534/module.mk
 include %reldir%/bug-51599/module.mk
+include %reldir%/bug-52075/module.mk
 include %reldir%/class-concat/module.mk
 include %reldir%/classdef/module.mk
 include %reldir%/classdef-multiple-inheritance/module.mk
 include %reldir%/classes/module.mk
 include %reldir%/ctor-vs-method/module.mk
 include %reldir%/fcn-handle-derived-resolution/module.mk
 include %reldir%/nest/module.mk
 include %reldir%/publish/module.mk
